File: audio/music-player/music-player.js
/* Plays music based on triggers
 *
 * To-Do:
 * - Rewrite script to C#
 */

#pragma strict
#pragma downcast

// Soundtrack list
var SoundtrackList : AudioClip[];

// Booleans
var FadeOutBetweenTracks : boolean = true;
var FadeInNewTrack : boolean = true;

// Speeds
var FadeOutSpeed : float = 1.0;
var FadeInSpeed : float = 1.0;

// Delay between tracks
var TransitionDelay : float = 1.0;

// Additional (hidden)
@HideInInspector var fadeOutBool : boolean = false;
@HideInInspector var fadeInBool : boolean = false;
@HideInInspector var fadeout_speed_fix : float = 1.0;
@HideInInspector var fadein_speed_fix : float = 1.0;
@HideInInspector var track_id : int;

function Start() {
	// Convert speed into actual seconds
	fadeout_speed_fix = 1 / FadeOutSpeed;
	fadein_speed_fix = 1 / FadeInSpeed;
}

function playTrack(id) {
	track_id = id;
	if(audio.isPlaying && FadeOutBetweenTracks) {
		fadeOutBool = true;
	} else {
		if(FadeInNewTrack) {
			audio.volume = 0;
			audio.clip = SoundtrackList[track_id];
			audio.Play();
			fadeIn();
		} else {
			audio.volume = 1;
			audio.clip = SoundtrackList[track_id];
			audio.Play();
		}
	}
}

function fadeOut() {
	if(!fadeOutBool) {
		audio.Stop();
		yield WaitForSeconds(TransitionDelay);
		playTrack(track_id);
	}
}

function fadeIn() {
	fadeInBool = true;
}

function Update() {
	if(fadeOutBool) {
		if(audio.volume != 0) {
			audio.volume -= 1 * Time.deltaTime * fadeout_speed_fix;
		} else {
			fadeOutBool = false;
			fadeOut();
		}
	}

	if(fadeInBool) {
		if(audio.volume != 1) {
			audio.volume += 1 * Time.deltaTime * fadein_speed_fix;
		} else {
			fadeInBool = false;
		}
	}
}

File: audio/music-player/MusicPlayer.cs
/* Plays music based on triggers
 */

using System.Collections;
using UnityEngine;



public class MusicPlayer : MonoBehaviour
{
	// Soundtrack list
	public AudioClip[] soundtrackList;
	
	// Booleans
	public bool fadeOutBetweenTracks = true;
	public bool fadeInNewTrack = true;
	
	// Speeds
	public float fadeOutSpeed = 1.0f;
	public float fadeInSpeed = 1.0f;
	
	// Delay between tracks
	public float transitionDelay = 1.0f;
	
	// Additional (private)
	bool _isFadingOut = false;
	bool _isFadingIn = false;
	float _fadeoutSeconds = 1.0f;
	float _fadeinSeconds = 1.0f;
	int _trackID;
	
	
	void Start()
	{
		// Convert speed into actual seconds
		_fadeoutSeconds = 1f / this.fadeOutSpeed;
		_fadeinSeconds = 1f / this.fadeInSpeed;
	}
	
	public void PlayTrack(int id)
	{
		AudioSource audio = this.audio;
		
		_trackID = id;
		if (audio.isPlaying && this.fadeOutBetweenTracks) {
			_isFadingOut = true;
		} else {
			if (this.fadeInNewTrack) {
				audio.volume = 0f;
				audio.clip = this.soundtrackList[_trackID];
				audio.Play();
				FadeIn();
			} else {
				audio.volume = 1f;
				audio.clip = this.soundtrackList[_trackID];
				audio.Play();
			}
		}
	}
	
	void FadeOut() {
		StartCoroutine(FadeOutCorountine());
	}
	IEnumerator FadeOutCorountine()
	{
		if (!_isFadingOut) {
			this.audio.Stop();
			yield return new WaitForSeconds(this.transitionDelay);
			PlayTrack(_trackID);
		}
	}
	
	void FadeIn()
	{
		_isFadingIn = true;
	}
	
	void Update()
	{
		AudioSource audio = this.audio;
		
		if (_isFadingOut) {
			if (audio.volume > 0f) {
				audio.volume -= Time.deltaTime * _fadeoutSeconds;
			} else {
				_isFadingOut = false;
				FadeOut();
			}
		}
	
		if (_isFadingIn) {
			if (audio.volume < 1f) {
				audio.volume += Time.deltaTime * _fadeinSeconds;
			} else {
				_isFadingIn = false;
			}
		}
	}
}
